packages/astro/src/webcontainer-files/cache.ts

```diff
@@ -0,0 +1,198 @@
+    this._resolve = resolve;
+
+    for (const filesMap of _filesMapGraph.allFilesMap()) {
+      this._requestsQueue.add(filesMap.path);
+      this._cache.set(getFilesRef(filesMap.path, this._dirs), undefined);
+    }
+
+    this._generateFileMaps();
+  }
+
+  generateFileMapForPath(filePath: string) {
+    const filesMapFolderPath = resolveFilesFolderPath(filePath, this._logger, this._dirs);
+
+    if (!filesMapFolderPath) {
+      this._logger.warn(`File ${filePath} is not part of the tutorial content or templates folders.`);
+      return;
+    }
+
+    const filesRef = getFilesRef(filesMapFolderPath, this._dirs);
+
+    // clear the existing cache value (mark it as stale)
+    this._cache.set(filesRef, undefined);
+
+    // if it's a template do the same for its dependencies recursively
+    this._invalidateCacheForDependencies(filesMapFolderPath);
+
+    // add this file as required to be processed
+    this._requestsQueue.add(filesMapFolderPath);
+
+    if (this._timeoutId) {
+      return;
+    }
+
+    const { promise, resolve } = withResolvers<void>();
+
+    this._readiness = promise;
+    this._resolve = resolve;
+    this._timeoutId = setTimeout(this._generateFileMaps, 100);
+  }
+
+  async canHandle(reqURL: string | undefined): Promise<string | false> {
+    const fileMapPath = new URL(reqURL ?? '/', 'http://a').pathname.slice(1);
+
+    // if it's not in the cache then this path is not handled by this cache
+    if (!this._cache.has(fileMapPath)) {
+      return false;
+    }
+
+    let cacheValue = this._cache.get(fileMapPath);
+
+    // if the value is not present the cache is not fresh
+    if (typeof cacheValue === 'undefined') {
+      await this._readiness;
+
+      cacheValue = this._cache.get(fileMapPath);
+    }
+
+    if (typeof cacheValue === 'undefined') {
+      this._logger.error(`The cache never resolved for ${fileMapPath}.`);
+      return false;
+    }
+
+    // mark the path as "hot", meaning we should refresh the files in webcontainer once the file generation has completed
+    this._hotPaths.add(fileMapPath);
+
+    return cacheValue;
+  }
+
+  private _invalidateCacheForDependencies(folderPath: string) {
+    const filesMap = this._filesMapGraph.getFilesMapByFolder(folderPath);
+
+    if (!filesMap) {
+      return;
+    }
+
+    for (const dependency of filesMap.allDependents()) {
+      this._cache.set(getFilesRef(dependency.path, this._dirs), undefined);
+    }
+  }
+
+  private _generateFileMaps = async () => {
+    const hotFilesRefs: string[] = [];
+
+    while (this._requestsQueue.size > 0) {
+      // first we update all files map
+      for (const folderPath of this._requestsQueue) {
+        this._filesMapGraph.updateFilesMapByFolder(folderPath, this._logger);
+      }
+
+      // then we make sure all their dependencies are added as well
+      for (const folderPath of this._requestsQueue) {
+        for (const dependency of this._filesMapGraph.getFilesMapByFolder(folderPath)!.allDependents()) {
```

Comment:
Should the falsy return values of `getFilesMapByFolder` be handled here instead? Or make `getFilesMapByFolder` return an object with dummy `.allDependents()` instead of `undefined`?

Reply:
It cannot return a falsy value right now because `updateFilesMapByFolder` called in the previous loop always adds the folder it received to the graph.

It could return a falsy value if `updateFilesMapByFolder` was modified in the future and would no longer provide that guarantee.

What about logging a `This is a bug!` message if the value is `undefined`? Wdyt?

Reply:
Oh right, it's just TS that doesn't necessary keep up with the types here. I think this is fine as is. In case of bug it will still throw on `.allDependents` access.